x/gamm/pool-models/stableswap/README.md

```diff
@@ -1,4 +1,106 @@
+
+    SpotPrice(baseAssetDenom string, quoteAssetDenom string) (sdk.Dec, error)
+
+	JoinPool(tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, err error)
+	JoinPoolNoSwap(tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, err error)
+	ExitPool(numShares sdk.Int, exitFee sdk.Dec) (exitedCoins sdk.Coins, err error)
+```
+
+The "constant" part of CFMM's imply that we can reason about all their necessary algorithms from just the CFMM equation. There are still multiple ways to solve each method. We detail below the ways in which we do so. This is organized by first discussing variable substitutions we do, to be in a more amenable form, and then the details of how we implement each method.
+
+<!---  This is true, but however the details of this curve actually lend itself better to using more general solutions for swaps via binary search. (Inspired from https://arxiv.org/abs/2111.13740 ) The remainder of the methods we implement via generic CFMM ideas.
+
+We detail below both what the direct solution via the CFMM equation is, and the solution we use.
+Then we show how these are used to give all of the swqp equations. --->
+
+### CFMM function
+
+Most operations we do only need to reason about two of the assets in a pool, and sometimes only one.
+We wish to have a simpler CFMM function to work with in these cases.
+Due to the CFMM equation $f$ being a symmetric function, we can wlog reorder the arguments to the function. Thus we put the assets of relevance at the beginning of the function. So if two assets $x, y$, we write: $f(x,y, a_3, ... a_n) = xy * a_3 * ... a_n (x^2 + y^2 + a_3^2 + ... + a_n^2)$.
+
+We then take a more convenient expression to work with, via variable substition.
+$$ \begin{equation}
+    v =
+    \begin{cases}
+      1, & \text{if}\ n=2 \\
+      \prod_{i=3}^n a_i, & \text{otherwise}
+    \end{cases}
+  \end{equation} \newline 
+  
+  \begin{equation}
+    w =
+    \begin{cases}
+      0, & \text{if}\ n=2 \\
+      \sum_{i=3}^n a_i^2, & \text{otherwise}
+    \end{cases}
+  \end{equation} \newline
+  \text{then } g(x,y,v,w) = xyv(x^2 + y^2 + w) = f(x,y, a_3, ... a_n) 
+$$
+
+### Swaps
+
+First notice that for all swaps, we only deal with two assets at a time, as swaps are given one asset in, and one asset out. For exposition, lets call the input asset $x$, the output asset $y$, and we can compute $v$ and $w$ given the other assets, whose reserves are untouched.
+
+First we note the direct way of solving this, its limitation, and then the binary search equations.
+
+#### Direct swap solution
+
+Suppose the existence of a function $\text{solve\_cfmm}(x, v, w, k) = y\text{ s.t. }g(x, y, v, w) = k$.
+Then we can solve swaps by first computing $k = g(x_0, y_0, v, w)$.
+Then suppose I want to swap $a$ units of $x$, and then want to find how many units $b$ of $y$ that we get out.
+We do this by computing $y_f = \text{solve\_cfmm}(x_0 + a, v, w, k)$, and then $b = y_0 - y_f$
```

Comment:
This and all of the remaining expressions also do not format for me unfortunately